model-engine/model_engine_server/inference/tool_completion/tools.py

```diff
@@ -0,0 +1,249 @@
+import re
+import subprocess
+from enum import Enum
+from typing import Optional, Tuple
+
+import docker
+from model_engine_server.inference.tool_completion.base import BaseTool
+from model_engine_server.inference.tool_completion.utils import (
+    FIX_ERRORS_MAPPING,
+    NAME_ERROR_PATTERN,
+    PRINT_PATTERN,
+)
+from transformers import LlamaTokenizer
+
+tokenizer = LlamaTokenizer.from_pretrained("openlm-research/open_llama_7b", legacy=False)
+MAX_CODEBLOCK_RETRIES = 3
+
+
+class CodeBlockEvaluator(BaseTool):  # pragma: no cover
+    """
+    A evaluator to "pseudo-safely" execute python code blocks.
+    Executes code from a model generated response using a safe python interpreter.
+    the code should have the following format:
+
+    ```python
+    {code}
+    ```
+    {output}
+    >>>
+
+    The output will be replaced with the output from executing the code.
+    """
+
+    tool_context_start = "```python\n"
+    tool_call_token = "\n```\n"
+    tool_context_end = "\n>>>\n"
+
+    @staticmethod
+    def _cleanup_code_error(error_code: str) -> str:
+        """This function will clean up an error code from code execution
+
+        Args:
+            error_code (str): The full error code (e.g. like below):
+
+            Command '['python', '-c', 'import math\nx = 2\nmath.sqrt(y)']' in image 'continuumio/anaconda3'
+              returned non-zero exit status 1: b'Traceback (most recent call last):
+              File "<string>", line 3, in <module>\nNameError: name \'y\' is not defined\n'
+
+        Returns:
+            str: like the following:
+
+            Traceback (most recent call last): File "<string>", line 3, in <module>
+            NameError: name \'y\' is not defined
+
+        """
+        if "Traceback" not in error_code:
+            return error_code
+
+        # Let's find the byte string: (e.g. b')
+        stacktrace = error_code.split("b'")[-1]
+
+        # Now read it as a bytestring
+        stacktrace = "\n" + stacktrace.encode("utf-8").decode("unicode_escape")
+
+        return stacktrace.strip("'")
+
+    def __init__(self):
+        # Condition to check if we can use docker
+        try:
+            self.client = docker.from_env()
+            self.evaluate = self.evaluate_code_in_docker
+        except docker.errors.DockerException:
+            # If docker is not available, use the python interpreter
+            self.evaluate = self.evaluate_code_using_exec
```

Comment:
is this safe for the batch job? seems that potentially harmful generations could break this

Reply:
(synced offline) known risk - for now since main use is internal, risk should be low and this is _okay_. will have follow-ups to mitigate issues

Reply:
that's true. please take a look how we think about this problem: https://docs.google.com/document/d/1zKfhizd0FpwoupF_jYmWkz4naTxkInI0g-rS6xzNWso/edit#heading=h.2hej6pszhyrd

Reply:
i think for batch jobs the risk is more manageable